### ISSUE TITLE
Error on inbound connection

### ISSUE THREAD

**Issue by carver** [ethereum/trinity#1895]
### What is wrong?

```wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x7f8028142c10>()]>>
   ERROR  2020-07-14 14:11:42,771            FullServer  Unexpected error handling handshake with ('195.14.184.226', 58872)
Traceback (most recent call last):
  File "/home/jcarver/code/trinity/trinity/server.py", line 135, in receive_handshake
    await self._receive_handshake(reader, writer)
  File "/home/jcarver/code/trinity/trinity/server.py", line 160, in _receive_handshake
    protocol_handshakers=handshakers,
RuntimeError: coroutine ignored GeneratorExit
   ERROR  2020-07-14 14:11:42,774               asyncio  Task exception was never retrieved
future: <Task finished coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('cannot reuse already awaited coroutine')>
Traceback (most recent call last):
  File "/home/jcarver/code/trinity/p2p/multiplexer.py", line 289, in _stream_protocol_messages
    yield await msg_queue.get()
RuntimeError: cannot reuse already awaited coroutine
```

### How can it be fixed

:man_shrugging:

**Comment by gsalgado** [comment]
This is weird; my node can receive incoming connections (at least from another locally running trinity instance). Can you try that to see if it works? Is this on master or on #1892?

**Comment by carver** [comment]
> Is this on master or on #1892?

It first showed up when I started with #1892 -- but then because UPnP was configured, even when I ran it without the patch I was getting the error. It seems to go away when I do this:
```patch
diff --git a/trinity/server.py b/trinity/server.py
index 9402862b..e006f03f 100644
--- a/trinity/server.py
+++ b/trinity/server.py
@@ -172,7 +172,10 @@ class BaseServer(Service, Generic[TPeerPool]):
             peer = factory.create_peer(connection)
 
             if self.peer_pool.is_full:
-                await peer.disconnect(DisconnectReason.TOO_MANY_PEERS)
+                try:
+                    await peer.disconnect(DisconnectReason.TOO_MANY_PEERS)
+                except AttributeError:
+                    pass
                 return
             elif not self.peer_pool.is_valid_connection_candidate(peer.remote):
                 await peer.disconnect(DisconnectReason.USELESS_PEER)
```
(weirdly the exception that inspired the patch was logged separately, later)

So I suspect it will actually get fixed by #1840

**Comment by carver** [comment]
Hm, actually there are more crashes related to `p2p_api` after I apply that patch, but still hopefully solved by #1840

**Comment by carver** [comment]
Nope, #1840 didn't fix it. Still seeing this *very* regularly. Running on 1a1589346e41d

**Comment by carver** [comment]
Also, this patch doesn't seem to help anymore: https://github.com/ethereum/trinity/issues/1895#issuecomment-658912733

**Comment by gsalgado** [comment]
@carver care to paste your full logs so that I can see if I find anything that will shed some light on what could be causing it?  Also, do you get that error if you connect two trinity instances locally?

**Comment by carver** [comment]
DM'd the logs. They were generated at 10a925079032b (master).

No apparent problem in a simple local trinity<->trinity test.

**Comment by gsalgado** [comment]
I just saw this as well, but got a different exception, probably because I'm on python3.8 and the one reported by @carver was on python3.7. Also, in my case I see 3 inbound connection attempts from the same host, all within a few milliseconds:

```
task: <Task pending name='Task-388401' coro=<BaseServer.receive_handshake() running at /home/salgado/src/snakecharmers/trinity/trinity/server.py:134> wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x7fc4c7dba910>()]>>
   ERROR  2020-08-25 15:52:03,019               asyncio  Task was destroyed but it is pending!
task: <Task pending name='Task-388402' coro=<BaseServer.receive_handshake() running at /home/salgado/src/snakecharmers/trinity/trinity/server.py:134> wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x7fc4cdc688e0>()]>>
   ERROR  2020-08-25 15:52:03,019               asyncio  Task was destroyed but it is pending!
task: <Task pending name='Task-388403' coro=<BaseServer.receive_handshake() running at /home/salgado/src/snakecharmers/trinity/trinity/server.py:134> wait_for=<_GatheringFuture pending cb=[<TaskWakeupMethWrapper object at 0x7fc4cdc686a0>()]>>
   ERROR  2020-08-25 15:52:03,019            FullServer  Unexpected error handling handshake with ('3.236.146.3', 46962)
Traceback (most recent call last):
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 134, in receive_handshake
    await self._receive_handshake(reader, writer)
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 154, in _receive_handshake
    connection = await receive_dial_in(
RuntimeError: coroutine ignored GeneratorExit
   ERROR  2020-08-25 15:52:03,023            FullServer  Unexpected error handling handshake with ('3.236.146.3', 46966)
Traceback (most recent call last):
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 134, in receive_handshake
    await self._receive_handshake(reader, writer)
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 154, in _receive_handshake
    connection = await receive_dial_in(
RuntimeError: coroutine ignored GeneratorExit
   ERROR  2020-08-25 15:52:03,024            FullServer  Unexpected error handling handshake with ('3.236.146.3', 46964)
Traceback (most recent call last):
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 134, in receive_handshake
    await self._receive_handshake(reader, writer)
  File "/home/salgado/src/snakecharmers/trinity/trinity/server.py", line 154, in _receive_handshake
    connection = await receive_dial_in(
RuntimeError: coroutine ignored GeneratorExit
 WARNING  2020-08-25 15:52:03,107       Sync / PeerPool  Event loop blocked or overloaded: delay=0.373s, tasks=579
   ERROR  2020-08-25 15:52:03,144               asyncio  Task exception was never retrieved
future: <Task finished name='Task-389958' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>
RuntimeError: aclose(): asynchronous generator is already running
   ERROR  2020-08-25 15:52:03,144               asyncio  Task exception was never retrieved
future: <Task finished name='Task-389959' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>
RuntimeError: aclose(): asynchronous generator is already running
   ERROR  2020-08-25 15:52:03,144               asyncio  Task exception was never retrieved
future: <Task finished name='Task-389960' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>
RuntimeError: aclose(): asynchronous generator is already running
```

**Comment by gsalgado** [comment]
So, I've tracked this down to the following code:

https://github.com/ethereum/trinity/blob/e87c30968b20b883ae825cee8a2c33b67c068ae2/p2p/handshake.py#L371-L382

For some reason that `negotiate_protocol_handshakes()` call is raising a `GeneratorExit`, but I couldn't figure out where it comes from, because the one raised when we leave the `multiplexer.stream_protocol_messages()` context in https://github.com/ethereum/trinity/blob/e87c30968b20b883ae825cee8a2c33b67c068ae2/trinity/protocol/eth/handshaker.py#L149 correctly causes the generator to terminate without bubbling up. That `GeneratorExit` is then caught on line 376 and we attempt to call `transport.close()`, which in turn seems to raise yet another `GeneratorExit`.

I've changed line 376 to re-raise the `GeneratorExit` without calling `transport.close()` and after that I no longer get the `RuntimeError: coroutine ignored GeneratorExit` errors, but I still get the `Task was destroyed but it is pending!` and `RuntimeError: aclose(): asynchronous generator is already running` ones.

As I can't find any other places (in the handshake code path) where we use an async generator, I went looking for python issues and found https://bugs.python.org/issue39116, which unfortunately hasn't received any feedback.

Although this is happening somewhat frequently, it is not actually crashing trinity and only happens on certain inbound connections, so I'm going to stop working on it for now as it's driving me crazy
